### DISCUSSION TITLE
Redux

### DISCUSSION THREAD

**Discussion by FunctionalMetatable** [ScratchAddons/ScratchAddons#3346] [Help] [answered]
Not sure if this is possible on all pages, but we could use `document.querySelector("#app").__reactContainer<HASH>.child.stateNode.store` to access internal redux state instead of replacing `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__` and using Reducks? i'm not sure about redux that much but they equal true in console
![image](https://user-images.githubusercontent.com/67440879/130356835-c6eae31f-397b-4ad4-9647-9ddc15330400.png)

**Comment by apple502j** [comment]
Current design works fine.

**Reply by FunctionalMetatable** [reply]
~~@apple502j But by using this method we could stop conflicts?~~ nvm we cant do reduxEvents

**Comment by TheColaber** [comment]
Whow thats really cool... To avoid the random $xxxx... at the end of the variable (which is different in every page), we can use this:
```js
document.querySelector("#app")._reactRootContainer._internalRoot.current.child.stateNode.store.getState()
```

and, for turbowarp support, this works too:
```js
let rrContainer = document.querySelector("#app")._reactRootContainer;
(rrContainer._internalRoot || rrContainer).current.child.stateNode.store.getState();
```

**Reply by TheColaber** [reply]
@apple502j doesn't this design look better than the current, if not why.

**Reply by WorldLanguages** [reply]
we can't dispatch events

**Reply by FunctionalMetatable** [reply]
> we can't dispatch events

we can, replace getState with dispatch

**Comment by apple502j** [comment, marked as answer]
Update: I think we can do this - redux events can stay. One thing to note is that Scratch is using older version of Redux (3.5.2 or 3.7.2 depending on code) so my solution might not work.
